src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -150,14 +186,81 @@ private TypeSymbolWithAnnotations LvalueResultType
+        private void SetInvalidResult() => SetResult(_invalidType, _invalidType.ToTypeSymbolWithAnnotations(), updateAnalyzedNullability: false);
+
+        private void SetResult(TypeWithState resultType, TypeSymbolWithAnnotations lvalueType, bool updateAnalyzedNullability = true, bool? isLvalue = null)
         {
             _visitResult = new VisitResult(resultType, lvalueType);
+            // We use the rvalue type for the public nullability because in most scenarios with differing lvalue and rvalue nullabilities,
+            // the public API cares more about the current state of the expression, not what can be assigned into it. Scenarios that care only
+            // about lvalue will use LvalueResultType, which will use the annotation for the nullable state instead.
+            if (updateAnalyzedNullability)
+            {
+                SetAnalyzedNullability(_currentExpression, _visitResult, isLvalue);
+            }
+        }
+
+        private bool ShouldMakeNotNullRvalue(BoundExpression node, bool wasReachable) => node.IsSuppressed || node.HasAnyErrors || !wasReachable;
+
+        /// <summary>
+        /// Sets the analyzed nullability of the expression to be the given result. droppingInfo is only used by UseLvalueOnly and UseRvalueOny, to
+        /// bypass the asserts for changing the l or r value of an already-set expression incorrectly.
+        /// </summary>
+        private void SetAnalyzedNullability(BoundExpression expr, VisitResult result, bool? isLvalue = null)
+        {
+            if (expr == null || _disableNullabilityAnalysis) return;
+
+#if DEBUG
+            // PROTOTYPE(nullable-api): detect when coming in from a bad expression and don't assert for those
+            //Debug.Assert(!s_skippedExpressions.Contains(expr.Kind), $"Should not be analyzing {expr}");
+            // PROTOTYPE(nullable-api): Does always seem to be holding true: this assert is essential for ensuring that we aren't
+            // changing the observable results of GetTypeInfo.
+            //Debug.Assert((result.RValueType.Type is null && result.LValueType.TypeSymbol is null && expr.Type is null) ||
+            //             result.RValueType.Type.Equals(expr.Type, TypeCompareKind.ConsiderEverything |
+            //                                                      TypeCompareKind.AllNullableIgnoreOptions |
+            //                                                      TypeCompareKind.IgnoreDynamicAndTupleNames |
+            //                                                      TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds),
+            //             $"Cannot change the type of {expr} from {expr.Type} to {result.RValueType.Type}");
+#endif
+
+            if (_analyzedNullabilityMapOpt != null)
+            {
+                // PROTOTYPE(nullable-api): enable and verify these assertions
+#if false
+                if (_analyzedNullabilityMapOpt.TryGetValue(expr, out var existing))
```

Comment:
>if (_analyzedNullabilityMapOpt.TryGetValue(expr, out var existing)) [](start = 16, length = 67)

Can we simply assert the same expression is not set twice?

Reply:
No. There are lots of times when an expression will be set twice, including the common path of `HasAnyErrors` in `VisitExpressionWithoutStackGuard`

---
In reply to: [265683897](https://github.com/dotnet/roslyn/pull/34017#discussion_r265683897) [](ancestors = 265683897)

Reply:
Would those cases (where we set the same expression state twice) disappear if we only called `SetAnalyzedNullability` in `VisitExpressionWithoutStackGuard`, where `_callbackOpt` was invoked previously?

---
In reply to: [265690916](https://github.com/dotnet/roslyn/pull/34017#discussion_r265690916) [](ancestors = 265690916,265683897)

Reply:
No. There are lots of scenarios where we don't use the standard visit logic (conversions and object initializers being the biggest, but not only, examples of this), and there's enough different ways that can happen that I would have no confidence that this would never happen.

---
In reply to: [266018238](https://github.com/dotnet/roslyn/pull/34017#discussion_r266018238) [](ancestors = 266018238,265690916,265683897)